bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/BufferedReadChannel.java

```diff
@@ -64,10 +71,22 @@ public int read(ByteBuf dest, long pos) throws IOException {
         return read(dest, pos, dest.writableBytes());
     }
 
+    @Override
+    public long size() throws IOException {
+        if (sealed) {
+            if (fileSize == -1) {
+                fileSize = validateAndGetFileChannel().size();
+            }
+            return fileSize;
+        } else {
+            return validateAndGetFileChannel().size();
+        }
+    }
```

Comment:
Would there be a benefit to override this logic in `BufferedChannel` class where the `position` (and possibly `unpersistedBytes`) field information would be leveraged? Doesn't `position` contain the file size when it's writable?

Reply:
The FileChnanel doesn't have `position` field and the `position()` method will trigger the OS system call

Reply:
I was referring to https://github.com/apache/bookkeeper/blob/22097343ceed1312964ad077cdcd9fbd7e3e026b/bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/BufferedChannel.java#L49-L52 in BufferedChannel class. In the case that the instance is a BufferedChannel class, perhaps it could use that information?

Is the challenge in that case that the same file could have a BufferedReadChannel instance and a BufferedChannel instance at the same time?

Reply:
I'm not got you fully, the BufferedChannel means the corresponding file is writing, it override the `BufferedReadChannel#read` method, it know the file end position.

> Is the challenge in that case that the same file could have a BufferedReadChannel instance and a BufferedChannel instance at the same time?

https://github.com/apache/bookkeeper/pull/4038#issuecomment-1652685967 
The test contains the case.